audio/out/ao.c

```diff
@@ -528,10 +528,13 @@ struct ao_device_list *ao_hotplug_get_device_list(struct ao_hotplug *hp,
             continue;
 
         if (ao->driver->hotplug_init) {
-            if (!hp->ao && ao->driver->hotplug_init(ao) >= 0)
-                hp->ao = ao; // keep this one
-            if (hp->ao && hp->ao->driver == d)
-                get_devices(hp->ao, list);
+            if (ao->driver->hotplug_init(ao) >= 0) {
+                get_devices(ao, list);
+                if (hp->ao)
+                    ao->driver->hotplug_uninit(ao);
+                else
+                    hp->ao = ao; // keep this one
+            }
```

Comment:
why does this call hotplug_unit on the existing hotplug ao?
shouldn't the code be more like this?
```c
            if (ao->driver->hotplug_init(ao) >= 0) {
                if (!hp->ao)
                    hp->ao = ao; // keep this one
                get_devices(hp->ao, list);
            }
```

Reply:
Only on AO will be saved in `ao_hotplug` and is the one expected to provide hotplug events.
The other ones are closed again.

Reply:
In that case I'd expect this:
```c
            if (ao->driver->hotplug_init(ao) >= 0) {
                get_devices(hp->ao, list);
                if (hp->ao) {
                    hp->ao->driver->hotplug_uninit(hp->ao);
                    talloc_free(hp->ao);
                }
                hp->ao = ao; // keep this one
            }
```

but wouldn't that override the choice of the initial if-block you added?

Reply:
Isn't this the same as the current code except for keep-last instead of keep-first?
Your proposal would indeed override the choice of the initial block. But the current code preserves it.

Reply:
Let's try a different approach because I still don't get it.
The commit message says "audio: list devices for all AOs with hotplug_init".
The following diff would satisfy this:
```diff
diff --git a/audio/out/ao.c b/audio/out/ao.c
index 4eb115194b..314db50e55 100644
--- a/audio/out/ao.c
+++ b/audio/out/ao.c
@@ -519,8 +519,7 @@ struct ao_device_list *ao_hotplug_get_device_list(struct ao_hotplug *hp)
         if (ao->driver->hotplug_init) {
             if (!hp->ao && ao->driver->hotplug_init(ao) >= 0)
                 hp->ao = ao; // keep this one
-            if (hp->ao && hp->ao->driver == d)
-                get_devices(hp->ao, list);
+            get_devices(ao, list);
         } else {
             get_devices(ao, list);
         }
```

what are the other changes for and why do we have them?

Reply:
Now `hotplug_init` is still only executed for one AO.

We would be executing `ao->list_devs()` (from `get_devices()`) without having executed `ao->hotplug_init()`.
The AO may not be written in a way that allows this because it differs from current behavior.